table/tables/partition.go

```diff
@@ -125,6 +129,43 @@ type PartitionExpr struct {
 	UpperBounds []expression.Expression
 	// Expr is the hash partition expression.
 	Expr expression.Expression
+
+	// The new range partition pruning
+	*ForRangePruning
```

Comment:
this is an optimization that avoids calculate `lessThenData` every time but doesn't exists in master branch?

Reply:
Yes, it's a TODO here https://github.com/pingcap/tidb/pull/15169/files#diff-0411f48b4045359c9f116ef6a779c3efR331

And another tiny change is lessThanData could be int/uint when I find a CI case fail.